### DISCUSSION TITLE
C++: why do we use CMake OBJECT library?

### DISCUSSION THREAD

**Discussion by KrisThielemans** [microsoft/yardl#189] [General]
The generated CMakeLists.txt uses an OBJECT library, e.g. see https://github.com/microsoft/yardl/blob/bf7068e93bb2115cbf731037ed1d75710d795679/cpp/evolution/v2/generated/CMakeLists.txt#L26C1-L32C2

This possibly avoids some issues, but it causes others, including the need for (unexpected) `target_include_directories`, as in
```cmake
set(PETSIRD_dir ../PETSIRD/cpp/generated)
add_subdirectory(${PETSIRD_dir} PETSIRD_generated)

add_executable(my_target blabla.cpp)
# without the following, generated include files are not found
target_include_directories(my_target PUBLIC ${PETSIRD_dir})
```

If OBJECT library is best, there probably should be a variable `<libraryname>_INCLUDE_DIRS`, which might be expected.

(Is the need for the `target_include_directories` documented?)

**Comment by johnstairs** [comment]
I think we can fix this in the object library library. I will take a look.

**Reply by KrisThielemans** [reply]
#104 asked for adding `target_include_directories`, but that doesn't seem to work for `OBJECT` libraries.

**Comment by johnstairs** [comment]
(And the idea with the object library is to make it easier to incorporate the generated code, but you are not required to use it)
